### PR TITLE
Adding ability to recurse a dist dir

### PR DESCRIPTION
**Pull request by jaustinpage**
This is another way to address https://github.com/gorilla-co/s3pypi/issues/88

### CODE REVIEW

File: s3pypi/core.py [hunk 1]
```diff
@@ -89,6 +90,18 @@ def parse_distribution(path: Path) -> Distribution:
     return Distribution(name, version, path)
 
 
+def parse_distributions(paths: List[Path]) -> List[Distribution]:
+    dists = []
+    for path in paths:
+        if path.is_file():
+            dists.append(parse_distribution(path))
+        if path.is_dir():
```
Comment (mdwint) [inline s3pypi/core.py:98]: Might as well make this `elif`.

File: s3pypi/core.py [hunk 2]
```diff
@@ -89,6 +90,18 @@ def parse_distribution(path: Path) -> Distribution:
...
+            for path_ in [f for f in path.iterdir() if f.is_file()]:
```
Comment (mdwint) [inline s3pypi/core.py:99]: You could also use round braces here to make this a generator rather than a list, which seems marginally better.

**Comment by jaustinpage** [comment]
I think being able to use tox and upload to s3 in a cross platform native python way is worth the extra lines of code. Let me know if you would like me to add documentation to how to use s3pypi in tox with example config, or if you would like to see better test coverage for this change,  or what other improvements I might be able to make.

**Review by mdwint — CHANGES_REQUESTED** [review]
I see how this could be beneficial in environments without shell expansion, but I'm also hesitant. What if someone writes `s3pypi foo/*` expecting this to go only one level deep, like most shell commands would? Am I being paranoid?

**Comment by jaustinpage** [comment]
> I see how this could be beneficial in environments without shell expansion, but I'm also hesitant. What if someone writes `s3pypi foo/*` expecting this to go only one level deep, like most shell commands would? Am I being paranoid?

This is a good point. Let me see if i can come up with a way to address shell expansion having already explored a directory in a clean way.

File: s3pypi/core.py (hunk 1 above)
Comment (jaustinpage) [reply s3pypi/core.py:98]: Yep, I considered this, and felt that the code was quicker to read this way. It does absolutely cost an extra path comparison though. I am happy to change it if you want.

File: s3pypi/core.py (hunk 2 above)
Comment (jaustinpage) [reply s3pypi/core.py:99]: I like generators. I tried to use a filter here, but, the way it was written didn't feel right, so i went with the list  comprehension. I forgot to see if I could go the generator route, good catch.

**Comment by jaustinpage** [comment]
> > I see how this could be beneficial in environments without shell expansion, but I'm also hesitant. What if someone writes `s3pypi foo/*` expecting this to go only one level deep, like most shell commands would? Am I being paranoid?
> 
> This is a good point. Let me see if i can come up with a way to address shell expansion having already explored a directory in a clean way.

**Comment by jaustinpage** [comment]
oops, accidentally closed this, mea culpa.

**Comment by jaustinpage** [comment]
> I see how this could be beneficial in environments without shell expansion, but I'm also hesitant. What if someone writes `s3pypi foo/*` expecting this to go only one level deep, like most shell commands would? Am I being paranoid?

we could add a --dir option, that allows you to specify a directory, and leave the existing args behavior as is.

**Comment by mdwint** [comment]
> we could add a --dir option, that allows you to specify a directory, and leave the existing args behavior as is.

That sounds reasonable. What do you think of calling it `-r, --recursive` (a boolean flag)?

**Comment by jaustinpage** [comment]
> > we could add a --dir option, that allows you to specify a directory, and leave the existing args behavior as is.
> 
> That sounds reasonable. What do you think of calling it `-r, --recursive` (a boolean flag)?

I am on board 100%. Ill get to work on it.

**Comment by jaustinpage** [comment]
> > > we could add a --dir option, that allows you to specify a directory, and leave the existing args behavior as is.
> > 
> > 
> > That sounds reasonable. What do you think of calling it `-r, --recursive` (a boolean flag)?
> 
> I am on board 100%. Ill get to work on it.

**Comment by jaustinpage** [comment]
ok, one of these days I am going to figure out how to not close the silly pr. very sorry for the email spam

**Comment by jaustinpage** [comment]
> > we could add a --dir option, that allows you to specify a directory, and leave the existing args behavior as is.
> 
> That sounds reasonable. What do you think of calling it `-r, --recursive` (a boolean flag)?

If we use a -r --recursive, should it go infinitely deep?

**Comment by mdwint** [comment]
> If we use a -r --recursive, should it go infinitely deep?

Sure, that's the most obvious behaviour given the name. I guess in practice deep recursion won't be needed much, and if I understand correctly you only need it for a single level, right?

**Comment by jaustinpage** [comment]
Hey, I had a realization. Because Tox is not auto-expanding the command line, we can just handle that specific case. We know that the file will come in as a non-existant file, since it is a file pattern. I am going to open a new pull request to look at this option and discuss.

**Comment by jaustinpage** [comment]
We should probably see if we like https://github.com/gorilla-co/s3pypi/pull/90 more than this pull request. If so, then we can close this one.

**Comment by jaustinpage** [comment]
closing in favor of #90

**Pull request closed without merging**